### PR TITLE
Fix tqdm

### PR DESCRIPTION
**Pull request by jeff-hernandez**
This PR refactors the progress bar.

### CODE REVIEW

File: composeml/label_maker.py [hunk 1]
```diff
@@ -106,16 +109,17 @@ def df_to_labels(df):
...
             bar_format += "Progress: {l_bar}{bar}| "
             bar_format += self.target_entity + ": {n}/{total} "
-            tqdm.pandas(bar_format=bar_format, ncols=90)
+            progress_bar = tqdm(total=labels.ngroups, bar_format=bar_format, ncols=90)
 
-        df = self._preprocess(df)
-        labels = df.groupby(self.target_entity)
-        apply = labels.progress_apply if verbose else labels.apply
-        labels = apply(df_to_labels, *args, **kwargs)
+        labels = labels.apply(df_to_labels, progress_bar=progress_bar, *args, **kwargs)
```
Comment (kmax12) [inline composeml/label_maker.py]: not part of this PR specifically, but do `*args, **kwargs` actually get passed through here? seems like `df_to_labels` doesn't take them?

File: composeml/label_maker.py [hunk 2]
```diff
@@ -106,16 +109,17 @@ def df_to_labels(df):
...
+        df = self._preprocess(df)
+        labels = df.groupby(self.target_entity)
+        progress_bar = None
+
         if verbose:
             bar_format = "Elapsed: {elapsed} | Remaining: {remaining} | "
             bar_format += "Progress: {l_bar}{bar}| "
             bar_format += self.target_entity + ": {n}/{total} "
-            tqdm.pandas(bar_format=bar_format, ncols=90)
+            progress_bar = tqdm(total=labels.ngroups, bar_format=bar_format, ncols=90)
```
Comment (kmax12) [inline composeml/label_maker.py]: do you need to set `ncols` yourself here? we don't do it in Featuretools, I don't have any issues I am aware of

Comment (kmax12) [inline composeml/label_maker.py]: rather than use `verbose`, you can pass the `disable=verbose` parameter to tqdm. that way it will automatically not show if verbose is false and you don't need special verbose logic

File: composeml/label_maker.py [hunk 3]
```diff
@@ -77,11 +77,14 @@ def search(self, df, minimum_data, num_examples_per_instance, gap, verbose=True,
...
-        def df_to_labels(df):
+        def df_to_labels(df, progress_bar=None):
+            labels = pd.Series(name=name)
             df = df.loc[df.index.notnull()]
             df.sort_index(inplace=True)
 
-            labels = pd.Series(name=name)
+            if progress_bar is not None:
+                not_done = progress_bar.n < progress_bar.total
+                progress_bar.update(n=int(not_done))
```
Comment (kmax12) [inline composeml/label_maker.py]: why does this get called if the progress bar is already completed?

Comment (kmax12) [reply composeml/label_maker.py]: should the update method get called at the end of this function onces it's actually completed?

also, we can update more fine grained after every iteration of `for example in range(num_examples_per_instance):`

basically `progress_bar.update(1/(ngroups * num_examples_per_instance))`

File: composeml/label_maker.py (hunk 1 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: That's correct, didn't see that. I removed them.

File: composeml/label_maker.py (hunk 2 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: When I don't use `ncols`, the progress bar stretches across the screen. Will send you examples.

Comment (jeff-hernandez) [reply composeml/label_maker.py]: Didn't know that. I made the update.

File: composeml/label_maker.py (hunk 3 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: That's correct. I fixed that. Also, change so that the progress bar for each example. Defined the total as `ngroups * num_examples_per_instance`.

File: composeml/label_maker.py [hunk 4]
```diff
@@ -101,21 +102,31 @@ def df_to_labels(df):
                     labels[cutoff_time] = label
 
                 cutoff_time = offset_time(df.index, gap)
+                not_done = progress_bar.n < progress_bar.total
```
Comment (kmax12) [inline composeml/label_maker.py]: why does this get called if the progress bar is already completed?

also, i think it's easier to read if we do

```
if progress_bar.n < progress_bar.total:
    progress_bar.update(n=1)
```

File: composeml/label_maker.py [hunk 5]
```diff
@@ -101,21 +102,31 @@ def df_to_labels(df):
...
+
+        progress_bar = tqdm(
+            total=total,
+            bar_format=bar_format,
+            disable=not verbose,
+            file=sys.stdout,
+        )
+
+        labels = labels.apply(df_to_labels, progress_bar=progress_bar)
+        n = progress_bar.total - progress_bar.n
```
Comment (kmax12) [inline composeml/label_maker.py]: why is this needed? shouldn't it be at total already?

File: composeml/label_maker.py (hunk 4 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: I thought the same, but with these adjustments:
```
-                not_done = progress_bar.n < progress_bar.total
-                progress_bar.update(n=int(not_done))
+                progress_bar.update(n=1)
```
Not sure why the progress bar renders this way.
```
150it [00:00, 316.63it/s]                                                                                 
```

File: composeml/label_maker.py (hunk 5 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: If the number of instances that the users sets is more than the number that actually exist in a group, the progress bar will not render completely.
```
labels = label_maker.search(
    df,
    minimum_data="1h",
    num_examples_per_instance=25,
    gap='30d',
    verbose=True,
)
```
```
Elapsed: 00:00 | Remaining: 00:00 | Progress:  14%|██████▌                                       | customer_id: 18/125 
```

Comment (kmax12) [reply composeml/label_maker.py]: we should account for that when you break early

```
if df.empty:
    progress_bar.update(n=skipped_iterations)
    break
```

File: composeml/label_maker.py (hunk 4 above)
Comment (kmax12) [reply composeml/label_maker.py]: why would `progress_bar.n < progress_bar.total` ever be false?

Comment (jeff-hernandez) [reply composeml/label_maker.py]: After a closer look, it seems like `GroupBy.apply` applies `df_to_labels` twice on the first group.

```
Elapsed: 00:00 | Remaining: ? | Progress:   0%|                      | customer_id: 0/125 

                     customer_id  amount
transaction_time                        
2014-01-01 03:13:51            1  120.52
2014-01-02 17:46:20            1   65.11
```

```
Elapsed: 00:00 | Remaining: 00:00 | Progress:  12%|█████▌           | customer_id: 15/125 

                     customer_id  amount
transaction_time                        
2014-01-01 03:13:51            1  120.52
2014-01-02 17:46:20            1   65.11

```
```
Elapsed: 00:00 | Remaining: 00:00 | Progress:  26%|███████████▊     | customer_id: 32/125 

                     customer_id  amount
transaction_time                        
2014-01-02 17:12:39            2  105.15
2014-01-04 05:45:43            2   51.95
```

As an alternative, maybe we can iterate the groups using a for loop?

Comment (kmax12) [reply composeml/label_maker.py]: ya, let's iterate over the groups then.

one note, I think they changed this in pandas `0.25.0`. even if it does work with the latest, i think we're better off having a lower minimum version. https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.25.0.html#groupby-apply-on-dataframe-evaluates-first-group-only-once

Comment (jeff-hernandez) [reply composeml/label_maker.py]: Made update to [iterate groups](https://github.com/FeatureLabs/compose-ml/pull/35/files#diff-22bf41f653d21f0ca3dd15dadf5e01e9R122).

File: composeml/label_maker.py (hunk 5 above)
Comment (jeff-hernandez) [reply composeml/label_maker.py]: Made update to [account for skipped iterations](https://github.com/FeatureLabs/compose-ml/pull/35/files#diff-22bf41f653d21f0ca3dd15dadf5e01e9R95).
